### ISSUE TITLE
hilbertCode is not really Hilbert SFC linear coordinate

### ISSUE THREAD

**Issue by oxidase** [Project-OSRM/osrm-backend#3352]
[hilbertCode](https://github.com/Project-OSRM/osrm-backend/blob/master/src/util/hilbert_value.cpp#L72-L82) returns a space filing curve linearizion, but it is not the Hilbert curve. 

Here is the recurrent motif of the current implementation 
<img src="https://cloud.githubusercontent.com/assets/4421046/20527523/18c3b458-b0c9-11e6-914f-a91cfbfc77f1.png" width="50">, but correct is 
<img src="https://cloud.githubusercontent.com/assets/4421046/20527545/2e4dad42-b0c9-11e6-955b-c506c2a87e92.png" width="50">.

For 5th recursion level of SFC linearizion looks  
<img src="https://cloud.githubusercontent.com/assets/4421046/20527646/71403d54-b0c9-11e6-99d0-6e56ac311415.png" width="160"> instead of 
<img src="https://cloud.githubusercontent.com/assets/4421046/20527656/7cf99316-b0c9-11e6-827e-57f2b69225d6.png" width="160">

It must be clarified  if the non-Hilbert SFC is the correct original decision.

Related PR  #3342

**Comment by danpat** [comment]
I can only think of two reasons:

  1. it's a bug (the function is `hilbertCode` after all).
  2.  the current implementation is an approximation that was fast and gave "good enough" locality for our purposes
  3. There is some subtle reason why an alternate curve shape performs better for OSRM's use-case.

My bet is on (1).  We can't know if it's (2) without talking to Dennis, and we can determine (3) by running some benchmarks.

Next step would be to profile https://github.com/Project-OSRM/osrm-backend/pull/3354 (I did some quick tests and at least in the benchmark scenario, performance was improved with the corrected curve).  If it performs better for every workload we can think of, let's make the change.

If it turns out that there is some magic performance bonus to the alternative motif, we should at least re-name the `hilbertCode` function call :-P

**Comment by TheMarex** [comment]
My money is on 1. This code never has test coverage and the result passes as "space-filling curve" which is really all the StaticRTree needs. Looking forward to the benchmark results, this could be great on big datasets.

**Comment by oxidase** [comment]
Even the Hilbert linearizion is some sort of heuristic and not the optimal packing. 
I have written a small python script that [dumps R-trees leaves](https://gist.github.com/oxidase/66ccac70893a582f6c968cc44b8fb996).
Here results for monaco-latest.osm [the current linearizion](https://gist.github.com/oxidase/1d0b0d4a26402c2a8a850dc4d366ae6e) and [the new one](https://gist.github.com/oxidase/a12f401f3fbe75f96ca8976e5001f9db)

Some normalized "area"-like value as (\sum leaf "areas")/ bb "area" changes from 1.76 to 0.97 it can be treated as an average number of touched leaves per query.  So in principle with the fix query time can be decreased.

**Comment by TheMarex** [comment]
@oxidase This looks much better. Great work. :+1: 

Side note: If we ever want to go optimal there is the STR packing algorithm [1] that yields optimal results on points (no overlap of leafs). I [implemented this for school once upon a time](https://github.com/alacarte-maps/alacarte/blob/master/include/general/rtree.hpp#L169-L191) but you can see that his code involves a lot of sorting. Not sure how well this scales on a planet scale dataset.

[1] http://www.dtic.mil/dtic/tr/fulltext/u2/a324493.pdf

**Comment by oxidase** [comment]
@TheMarex STR looks nice -> it will always involve only one leaf node in the search. One thing that has to be checked is how R-tree is balanced on the whole planet.

Anyway i found #2287 that must be fixed first. For example, source and target phantom nodes for "a" and "d" in features/car/traffic_turn_penalties.feature:105 change from 
```
input_coordinate = (lon:1.000899, lat:1)
   0,3 0 "ad"
input_coordinate = (lon:1.000899, lat:0.9991)
   0,3 0 "ad"
   3,4 0 "dhk"
   3,7 0 "def"
   2,3 0 "cd"
```
to 
```
input_coordinate = (lon:1.000899, lat:1)
   0,3 0 "ad"
input_coordinate = (lon:1.000899, lat:0.9991)
   3,4 0 "dhk"
   2,3 0 "cd"
   3,7 0 "def"
   0,3 0 "ad"
```
So instead of "ad"-"ad" routing finds "ad"-"dhk" that includes the negative turn turn penalty and  in the test including turning left is the fastest route.

**Comment by TheMarex** [comment]
@oxidase Do you think we can work around #2287 by fixing tests that rely on the order in which segments are picked? If that is not an option I would opt for marking them as broken with `@todo` and linking the mentioned issue.

I think this would be great for the 5.5 release if we can merge it in time.

**Comment by oxidase** [comment]
@TheMarex I have update PR. The only test that is unclear what to do is features/testbot/bearing_param.feature:48 - removing `ha` edge removes tie and test passes, but i don't know the original idea of edges `ha`, `dg`, `fc` and `be`, so just commented  out the row

@danpat benchmark for germany-latest shows change from 
```
Running raw RTree queries (1 result) with 10000 coordinates: Took 1.51699 seconds (1516.99ms)  ->  0.151699 ms/query (1516.99ms)
Running raw RTree queries (10 results) with 10000 coordinates: Took 1.63287 seconds (1632.87ms)  ->  0.163287 ms/query (1632.87ms)
```
to
```
Running raw RTree queries (1 result) with 10000 coordinates: Took 1.07445 seconds (1074.45ms)  ->  0.107445 ms/query (1074.45ms)
Running raw RTree queries (10 results) with 10000 coordinates: Took 1.14488 seconds (1144.88ms)  ->  0.114488 ms/query (1144.88ms)
 ```

**Comment by TheMarex** [comment]
This was fixed by @oxidase PR and will ship with 5.5.0 :tada: :tada:
